corva/models/stream.py

```diff
@@ -71,16 +74,55 @@ def app_stream_id(self) -> int:
         return self.metadata.app_stream_id
 
     @property
-    def is_completed(self) -> bool:
+    def _is_completed(self) -> bool:
         """there can only be 1 completed record, always located at the end of the list"""
 
         if self.records:
             return self.records[-1].collection == 'wits.completed'
 
         return False
 
+    @pydantic.root_validator(pre=False, skip_on_failure=True)
+    def require_app_key_in_metadata_apps(cls, values):
+        if values['app_key'] not in values['metadata'].apps:
+            raise ValueError('metadata.apps dict must contain an app key.')
+
+        return values
+
+    @pydantic.root_validator(pre=True)
+    def set_asset_id(cls, values):
+        """Dynamically sets value for asset_id.
+
+        asset_id could've been defined as property like below.
+
+        @property
+        def asset_id(self) -> Optional[int]:
+            return self.records[0].asset_id if self.records else None
+
+        The issue with the above method is:
+          after filtering, we may end up with empty records. Which leads to asset_id becoming None.
+          Using this validator we are able to dynamically set and store value of asset_id,
+          no matter what happens to records.
+        """
+
+        if 'asset_id' in values:
+            raise ValueError(
+                'asset_id can\'t be set manually, it is extracted from records automatically.'
```

Comment:
I don't think we need to be allergic to double quotes to surround the string here.